optuna/integration/pytorch_distributed.py

```diff
@@ -0,0 +1,249 @@
+
+        return _call_and_communicate(func, torch.float)
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> float:
+        def func() -> float:
+
+            assert self.delegate is not None
+            return self.delegate.suggest_uniform(name, low, high)
+
+        return _call_and_communicate(func, torch.float)
+
+    def suggest_loguniform(self, name: str, low: float, high: float) -> float:
+        def func() -> float:
+
+            assert self.delegate is not None
+            return self.delegate.suggest_loguniform(name, low, high)
+
+        return _call_and_communicate(func, torch.float)
+
+    def suggest_discrete_uniform(self, name: str, low: float, high: float, q: float) -> float:
+        def func() -> float:
+
+            assert self.delegate is not None
+            return self.delegate.suggest_discrete_uniform(name, low, high, q=q)
+
+        return _call_and_communicate(func, torch.float)
+
+    def suggest_int(self, name: str, low: int, high: int, step: int = 1, log: bool = False) -> int:
+        def func() -> float:
+
+            assert self.delegate is not None
+            return self.delegate.suggest_int(name, low, high, step=step, log=log)
+
+        return _call_and_communicate(func, torch.int)
+
+    def suggest_categorical(self, name: str, choices: Sequence["CategoricalChoiceType"]) -> Any:
+        def func() -> CategoricalChoiceType:
+
+            assert self.delegate is not None
+            return self.delegate.suggest_categorical(name, choices)
+
+        return _call_and_communicate_obj(func)
+
+    def report(self, value: float, step: int) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.report(value, step)
+        dist.barrier()
+
+    def should_prune(self) -> bool:
+        def func() -> bool:
+
+            assert self.delegate is not None
+            # Some pruners return numpy.bool_, which is incompatible with bool.
+            return bool(self.delegate.should_prune())
+
+        # torch.bool seems to be the correct type, but the communication fails
+        # due to the RuntimeError.
+        return _call_and_communicate(func, torch.uint8)
+
+    def set_user_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.set_user_attr(key, value)
+        dist.barrier()
+
+    def set_system_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.set_system_attr(key, value)
+        dist.barrier()
+
+    @property
+    def number(self) -> int:
+        def func() -> int:
+
+            assert self.delegate is not None
+            return self.delegate.number
+
+        return _call_and_communicate(func, torch.int)
+
+    @property
+    def trial_id(self) -> int:
```

Comment:
Do we still need to provide this API?

Reply:
I thought it was required by `BaseTrial`, but it wasn't. I removed it in https://github.com/optuna/optuna/pull/2303/commits/ac0002c1d21a058b6390a07c6c1c66c3be36f826. Thank you.